Paths: widgets.py record-editor.py

swagneto opened the pull request:
Topic: ArtistSelector Widget Addition
Description:
- Added the ArtistSelector widget to widgets.py, which combines a
  combobox and selection view, both backed by an artist model, to
  allow for easy search and selection of artists who contributed in
  some way to a particular art record.
- Integrated use of ArtistSelector into ArtRecordEditor.
- Fixed a bug in RubberBandedLabel.resizeEvent() where an old attribute
  name was still being referenced.

sdcolli1 committed:
Paths: widgets.py record-editor.py
Topic: ArtistSelector Widget Addition
Description:
  * Added the ArtistSelector widget to widgets.py, which combines a
    combobox and selection view, both backed by an artist model, to
    allow for easy search and selection of artists who contributed in
    some way to a particular art record.
  * Integrated use of ArtistSelector into ArtRecordEditor.
  * Fixed a bug in RubberBandedLabel.resizeEvent() where an old attribute
    name was still being referenced.

ghost commented:
Sorry this took a while, its been dog-and-pony-show week for our customer at work. Anywho, in response to your last email I changed the text color for new artists to red and fixed the focus propagation issue so your accelerators should work again. What I did **not** do was remove the new artist dialog/menu entry from ArtRecordEditor even though I agree that it's been obviated by ArtistSelector. I figured I'd give you the final say on that, but for future reference the QInputDialog class provides some handy static methods which basically provide everything that your LineEditDialog class does, but as a one-liner. For example, check out http://doc.qt.io/qt-5/qinputdialog.html#getText.

gthomsen commented:
This is sweet!  I played with this a bit last night and this morning and it dramatically speeds up cataloging.

There are a couple of minor things that are easy to fix:
1. The selection list shouldn't keep things sorted, but rather maintain the order artists are added.  Removing `self.selection_list.setSortingEnabled( True )` in the constructor does the trick.
2. Highlighting the combo box's contents once an artist is added to the selection list.  This lets you do: artist1<RET>artist2<RET> and get "artist1", "artist2" instead of "artist1", "artist1artist2".  Right now I need to to artist1<RET>C-u artist2<RET> to get that.

I'm not sure why though for some reason the auto-complete doesn't always kick in.  I can type a unique prefix (not an exact match) and get the dropdown to show the artist of interest though hitting RET inserts a new artist instead of adding the auto-completed one.  I noticed this with the "Unknown" (tried "unk", "Unk", and "Unknow"), "Grunf" (tried "grun" and "grunf"), and "Horde ZLA" (tried "hord", "horde zl", and "Horde") and all inserted new artists instead of auto-completing.

Exposed by this amazing widget is how ArtRecord's are created.  Right now they have a default artist of "Unknown" though that means the first action of a new record is usually removing "Unknown" and then adding something else.  I'm not sure if it's best to change the behavior in the Database module or in the editor itself - I need to think about this some more.

ghost commented:
Here's a patch which I think addresses some of issues above. Namely, I changed the completer of the combobox back to an inline mode, rather than a dropdown. That should match the behavior you're expecting, but now you won't be able to see _all_ potential matches at once, and if you need to add a new artist whose name happens to be the prefix of an existing artist, you'll need to take care to delete the auto-completed portion first. If this works for you I'll yank the extraneous code from using the dropdown completer and push up the changes.

[artist_selector_update.patch.txt](https://github.com/gthomsen/streetart-analysis/files/552264/artist_selector_update.patch.txt)

PS any thoughts on how you'd like to handle Unknown propagating to selector list yet?

gthomsen commented:
I haven't had a chance to think about the Unknown yet.  Here is what I was
thinking so you can add your $0.02 if you like:

Removing Unknown from ArtRecord:
Simplifies the editor's logic at the expense of making ArtRecord more
surprising.  ArtRecord's should always have at least one artist, and
relaxing things to allow zero means complicating the users of the module
with checking for correctness.  Even if this was removed from the module,
record-editor.py (and any other editor) would have to be modified to add
Unknown back in before committing changes.

Handling Unknown in the Editor:
Would keep ArtRecord's logic sensible though would require manipulation of
the records during the edit process - namely removing Unknown from the
artists list (only) when a new record is created, and then adding it to any
modified record if the artist list is empty.  This seems more sensible as
it is the editor that wants the records in a particular state to match the
editor's provided behavior.  Different editors could be fine with a default
artist.

I guess writing that out has convinced me the editor needs to be updated.

The patch works more inline with what I was describing, though after using
it for a few hours it seems to have some quirks to it.

Distinct cases autocomplete doesn't behave sensibly are:
- Target name is the same length as its lexiographic predecessor.  e.g.
  Seuf and Seud.
- Target name is shorter than its lexiographic predecessor.  e.g. Mark and
  Mario, Dip and Diaz, Menz and Mener.
- Target name is an extension of its lexiographic predecessor.  e.g. Roku
  and Rok.

Typing a full name that isn't autocompleted inserts what was typed, not
what should have been auto completed.  E.g. rude inserts "rude" instead of
autocompleting "Rude".  This happens quite a bit for shorter names (e.g.
BMS, NB) where they have to be typed case sensitively otherwise it inserts
something new.  Strangely, once I hit enter to autocomplete the completer
changes the typed contents to what I wanted ("bms RET" inserts "bms",
autocompletes "BMS") so hitting enter again would insert the desired artist.

WRT the combobox being replaced by the completer, I noticed that when I
type a prefix and attempt to scroll the combo box, it starts from the
beginning, rather than from where said prefix last matched an existing
entry (e.g. "Roku" click dropdown would put me at "Rok", assuming "Roku"
wasn't already in the database).

On Tue, Oct 25, 2016 at 10:16 PM, swagneto notifications@github.com wrote:

> Here's a patch which I think addresses some of issues above. Namely, I
> changed the completer of the combobox back to an inline mode, rather than a
> dropdown. That should match the behavior you're expecting, but now you
> won't be able to see _all_ potential matches at once, and if you need to
> add a new artist whose name happens to be the prefix of an existing artist,
> you'll need to take care to delete the auto-completed portion first. If
> this works for you I'll yank the extraneous code from using the dropdown
> completer and push up the changes.
> 
> artist_selector_update.patch.txt
> https://github.com/gthomsen/streetart-analysis/files/552264/artist_selector_update.patch.txt
> 
> PS any thoughts on how you'd like to handle Unknown propagating to
> selector list yet?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/gthomsen/streetart-analysis/pull/3#issuecomment-256239228,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AJjQYz6TiGFvjwSybUyZ7iorcdl9xVzfks5q3sYhgaJpZM4KaiMx
> .

ghost commented:
Round 2: this should address some, but not all of the concerns you laid out.

What this patch does:
- Address 'Unknown' artist within ArtistSelector widget. The literal string 'Unknown' is filtered from the list of provided artists to ArtistSelector, and if an ArtistSelector were to return an empty list of artists, ['Unknown'] is returned instead. Unknown is still available from the drop-down list, but selecting or entering it is essentially a no-op.
- Added a setting for the completer which should let it do binary search over the provided model. Just something in the API I noticed that I thought'd be a nice-to-have.
- Removed sorting on selection list.
- Fixed entered artists getting resolved to the 'canonical' form when auto-completing. I had accidentally commented out a bit too much in that first patch which caused the original issue (rude inserts "rude" instead of "Rude"), but I've since improved it a bit more. Basically, if you enter a new artist, whatever you provide will become the 'canonical' autocompleted text once you commit the record. However, while the artist name is still 'new' (i.e. red colored), you can retype the same name with a different capital/lowercase combination and it should update within the list. Just remember that once you commit the record, you're stuck with it unless you modify the artist model itself.
- The combobox line edit is now fully highlighted after hitting return, as requested.

What this latest patch does not do:
-  Address the 'non-sensible' behavior you outlined in the 3 bullet points. As far as I can tell, all those behaviors are baked into the 'inline' style completion (which is why I went with the popup style to start with). Might be possible to create a custom QCompleter widget which somehow gives you the best of both worlds, but I'd prefer to throw that to the backlog, so to speak.
- Address the autocomplete not updating the combobox index (the last finding). I experimented with this a little, but unfortunately updating the combobox current index triggers a slot that tries to update the line edit with the corresponding text, which is not what you want while you're also trying to type into it. I may play around with this more, but I didn't want it to hold up another update.

[artist_selector_update2.patch.txt](https://github.com/gthomsen/streetart-analysis/files/565635/artist_selector_update2.patch.txt)

sdcolli1 committed:
Paths: widgets.py
Topic: ArtistSelector Widget PR Updates
Description:
  * Made the following updates to address findings for the ArtistSelector
    pull request:
    - Addresses 'Unknown' artist within ArtistSelector widget. The literal
      string 'Unknown' is filtered from the list of provided artists to
      ArtistSelector, and if an ArtistSelector were to return an empty list
      of artists, ['Unknown'] is returned instead.
    - Added a setting for the completer which lets it do binary search over
      the provided model.
    - Removed sorting on selection list.
    - Fixed entered artists getting resolved to the 'canonical' form when
      auto-completing.
    - The combobox line edit is now fully highlighted after hitting return
      to streamline workflow.

ghost commented:
I went ahead and pushed up the changes I referred to in my last comment since that's probably easier than applying a patch. Also didn't want to forget about it as I wrap up work on the graphics framework transition.
